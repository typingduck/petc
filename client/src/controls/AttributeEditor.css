
.petc-attr-editor {
  background-color:  var(--editor-bg);
  border: var(--modal-border);
  border-radius: var(--modal-border-radius);
  box-shadow: var(--modal-box-shadow);
  padding: 20px 20px 20px;
  position: absolute;
  z-index: var(--z-index-modal);
}

.petc-attr-editor input {
  background-color: var(--editor-bg);
  border-radius: 5px;
  border: 1px solid rgba(210, 210, 210, 1);
  color: grey;
  font-family: var(--font);
  font-size: larger;
  font-weight: bold;
  text-align: center;
}

.petc-attr-editor td {
  text-align: center;
}

.petc-attr-editor option,
.petc-attr-editor select {
  background-color: var(--editor-bg);
  border-radius: 5px;
  border: 1px solid rgba(210, 210, 210, 1);
  color: grey;
  font-family: var(--font);
  font-size: larger;
  font-weight: bold;
  text-align: center;
  width: 100%;
  text-indent: 30%;

  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

