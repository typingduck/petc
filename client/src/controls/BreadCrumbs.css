
#petc-breadcrumbs {
  border-color: rgba(0, 0, 0, 0.2);
  border-radius: 10px;
  color: rgba(0, 0, 0, 0.2);
  font-size: larger;
  left: 30%;
  padding: 10px;
  position: absolute;
  text-align: center;
  top: 30%;
  z-index: var(--z-index-page-bg);
}
