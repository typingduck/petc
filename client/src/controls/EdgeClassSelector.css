#petc-edge-select-control {
  background-color: var(--modal-bg);
  border-radius: var(--modal-border-radius);
  border: var(--modal-border);
  border-right-width: var(--modal-border-wide);
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  border-left: none;
  bottom: 50px;
  box-shadow: var(--modal-box-shadow);
  left: 0;
  padding: 0;
  position: absolute;
  top: 20px;
  z-index: var(--z-index-modal);
}

#petc-edge-select-control ul {
  list-style: none;
  margin: 0;
  padding: 0;
}

#petc-edge-select-control li {
  align-items: center;
  border-bottom: 1px solid #555;
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-top: 2px;
  opacity: 0.8;
  padding: 10px;
}

#petc-edge-select-control li.selected {
  background-color: white;
}

.petc-edge-select-element {
  border: 1px solid rgba(0, 0, 0, 0);
  border-radius: 5px;
  height: 15px;
  position: relative;
  text-align: center;
  width: 80px;
}

.petc-edge-select-node-a {
  height: 10px;
  left: 0;
  position: absolute;
  width: 10px;
}

.petc-edge-select-node-b {
  height: 10px;
  position: absolute;
  right: 0;
  width: 10px;
}
