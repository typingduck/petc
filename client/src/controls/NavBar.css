
#petc-view-mode-control {
  background-color: var(--modal-bg);
  border: var(--modal-border);
  border-radius: var(--modal-border-radius);
  border-top-left-radius: 0;
  border-top-right-radius: 0;
  border-bottom-width: var(--modal-border-wide);
  border-top: none;
  box-shadow: var(--modal-box-shadow);
  font-size: 25px;
  overflow: hidden;
  position: absolute;
  right: 50px;
  top: 0;
  z-index: var(--z-index-modal);
}

#petc-view-mode-control ul {
  list-style: none;
  margin: 0;
  padding: 0;
}

#petc-view-mode-control li {
}

#petc-view-mode-control span {
  display: none;
}

#petc-view-mode-control a {
  color: grey;
}

#petc-view-mode-control li {
  display: inline-block;
  padding: 7px;
  border-right: 1px solid #555;
}

#petc-view-mode-control li.active {
  background-color: white;
}

#petc-view-mode-control li.active a {
}
