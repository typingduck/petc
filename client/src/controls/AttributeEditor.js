import React from 'react'
import Modal from './Modal'

import './AttributeEditor.css'
import {DEFAULT_CLASS_NAME} from '../model/model'

/**
 * Modal to edit label/class of edges or nodes.
 * A temporary solution using a dialog. Would be much better to edit in place
 * to keep the look and feel of a whiteboard but that will require a lot
 * more client side code to enable the feeling.
 */
class AttributeEditor extends React.Component {
  constructor (props) {
    super(props)

    this.state = {

      openEditor: false,
      label: null,
      className: null,
      classes: []
    }

    this.update = this.update.bind(this)
    this.apply = this.apply.bind(this)
    this.cancel = this.cancel.bind(this)
  }

  componentWillReceiveProps (nextProps) {
    const editAttributes = nextProps.controls.editAttributes
    const openEditor = Boolean(editAttributes)
    if (openEditor) {
      const itemId = nextProps.controls.editAttributes.id
      let toEdit, classes
      if (nextProps.doc.nodes[itemId]) {
        toEdit = nextProps.doc.nodes[itemId]
        classes = Object.keys(nextProps.doc.style.nodes)
      } else if (nextProps.doc.edges[itemId]) {
        toEdit = nextProps.doc.edges[itemId]
        classes = Object.keys(nextProps.doc.style.edges)
      } else {
        return
      }
      if (!classes.includes(DEFAULT_CLASS_NAME)) {
        classes = [DEFAULT_CLASS_NAME].concat(classes)
      }
      this.setState({
        openEditor: openEditor,
        label: toEdit.label || '',
        classes: classes,
        className: toEdit.className || DEFAULT_CLASS_NAME,
        pageX: editAttributes.pageX + 5,
        pageY: editAttributes.pageY + 5
      })
    } else {
      this.setState({ openEditor: false })
    }
  }

  update (e) {
    this._src = e.updated_src
  }

  onLabelChange (value) {
    this.setState({label: value})
  }

  onClassChange (value) {
    this.setState({className: value})
  }

  apply () {
    const itemId = this.props.controls.editAttributes.id
    if (getNode(this.props, itemId)) {
      const node = Object.assign({}, getNode(this.props, itemId))
      node.label = this.state.label
      node.className = this.state.className
      this.props.updateNode(node)
    } else if (getEdge(this.props, itemId)) {
      const edge = Object.assign({}, getEdge(this.props, itemId))
      edge.label = this.state.label
      edge.className = this.state.className
      this.props.updateEdge(edge)
    }
    this.cancel()
  }

  cancelNode () {
    this.setState({label: null, className: null})
    this.props.onCancel()
  }

  cancel () {
    this.props.setEditAttributes(null)
  }

  render () {
    return (
      <Modal
        className='petc-attr-editor'
        isOpen={this.state.openEditor}
        onRequestClose={this.cancel}
        pageX={this.state.pageX}
        pageY={this.state.pageY}
      >
        <table><tbody>
          <tr>
            <td>label:</td>
            <td>
              <input
                type='text'
                onChange={(ev) => this.onLabelChange(ev.target.value)}
                value={this.state.label}
              />
            </td>
          </tr>
          <tr>
            <td>class:</td>
            <td>
              <select
                value={this.state.className}
                onChange={(ev) => this.onClassChange(ev.target.value)}
              >
                {this.state.classes.map(className =>
                  <option key={className}>{className}</option>
                )}
              </select>
            </td>
          </tr>
        </tbody></table>
        <div style={{textAlign: 'center'}}>
          <button onClick={this.apply}>apply</button>
          <button onClick={this.cancel}>cancel</button>
        </div>
      </Modal>
    )
  }
}

function getNode (props, nodeId) {
  return props.doc.nodes[nodeId]
}

function getEdge (props, edgeId) {
  return props.doc.edges[edgeId]
}

export default AttributeEditor
