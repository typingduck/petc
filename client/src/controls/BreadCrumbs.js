import React from 'react'
import './BreadCrumbs.css'

/**
 * A useful development tool for viewing the redux state in the browser.
 * Hidden during normal viewing.
 */
export default function BreadCrumbs (props) {
  let msg = null

  if (props.controls.isNodeMode && Object.keys(props.doc.nodes).length === 0) {
    msg = <p> click on the canvas <br /> to create nodes </p>
  } else if (props.controls.isEdgeMode && Object.keys(props.doc.nodes).length < 2) {
    msg = <p> use the node view <br /> to create 2 or more nodes </p>
  } else if (props.controls.isEdgeMode && Object.keys(props.doc.edges).length === 0) {
    msg = <p> click on nodes <br /> to connect them </p>
  }

  return <div id='petc-breadcrumbs'> {msg} </div>
}
