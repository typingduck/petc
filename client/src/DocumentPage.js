import React from 'react'

import Kanvas from './kanvas/Kanvas'
import NavBar from './controls/NavBar'
import NodeClassSelector from './controls/NodeClassSelector'
import EdgeClassSelector from './controls/EdgeClassSelector'
import StoreView from './controls/StoreView'
import Trashcan from './controls/Trashcan'
import {loadDoc, connectSocket} from './model/model'

/**
 * Document view.
 * Wraps Kanvas with controls and connects up the networking.
 */
export default class DocumentPage extends React.Component {
  constructor (props) {
    super(props)
    this.socket = null

    this.addNode = this.addNode.bind(this)
    this.addNodeClass = this.addNodeClass.bind(this)
    this.updateNode = this.updateNode.bind(this)
    this.removeNode = this.removeNode.bind(this)
    this.addEdge = this.addEdge.bind(this)
    this.removeEdge = this.removeEdge.bind(this)
    this.addEdgeClass = this.addEdgeClass.bind(this)
    this.setDoc = this.setDoc.bind(this)
    this.receiveUpdate = this.receiveUpdate.bind(this)
    this.setSocket = this.setSocket.bind(this)
    this.networkFail = this.networkFail.bind(this)
  }

  componentDidMount () {
    connectSocket(this.props.docId, this.receiveUpdate)
      .then(this.setSocket)
      .catch(this.networkFail)
    loadDoc(this.props.docId)
      .then(this.setDoc)
      .catch(this.networkFail)
  }

  receiveUpdate (patch) {
    this.props.applyPatch(patch)
  }

  networkFail (err) {
    // TODO: handle
    /* eslint-disable no-console */
    /* globals console */
    console.error('network err: ', err)
    /* eslint-enable no-console */
  }

  setDoc (doc) {
    this.props.docLoaded(doc)
  }

  setSocket (socket) {
    this.socket = socket
  }

  addNode (node) {
    this.props.addNode(node)
    this.socket.addNode(node)
  }

  addNodeClass (name, style) {
    this.props.addNodeClass(name, style)
    this.socket.addNodeClass(name, style)
  }

  updateNode (node) {
    this.props.updateNode(node)
    this.socket.updateNode(node)
  }

  removeNode (nodeId) {
    this.props.removeNode(nodeId)
    this.socket.removeNode(nodeId)
  }

  addEdge (edge) {
    this.props.addEdge(edge)
    this.socket.addEdge(edge)
  }

  removeEdge (edgeId) {
    this.props.removeEdge(edgeId)
    this.socket.removeEdge(edgeId)
  }

  addEdgeClass (name, style) {
    this.props.addEdgeClass(name, style)
    this.socket.addEdgeClass(name, style)
  }

  render () {
    return (<div>
      <NavBar {...this.props} />
      <div className='petc-pageview'>
        <h1>pannus et circulos</h1>
        <Kanvas
          {...this.props}
          addNode={this.addNode}
          removeNode={this.removeNode}
          updateNode={this.updateNode}
          addEdge={this.addEdge}
          removeEdge={this.removeEdge}
        />
        <Trashcan {...this.props} />
        <NodeClassSelector
          {...this.props}
          addNode={this.addNode}
          addNodeClass={this.addNodeClass}
          addEdge={this.addEdge}
          updateNode={this.updateNode}
          removeNode={this.removeNode}
        />
        <EdgeClassSelector
          {...this.props}
          addEdgeClass={this.addEdgeClass}
        />
      </div>
      <StoreView {...this.props} />
    </div>)
  }
}
