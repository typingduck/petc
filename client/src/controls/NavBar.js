import React from 'react'
import {MdRemoveRedEye, MdGames, MdDeviceHub, MdShuffle, MdCreateNewFolder} from 'react-icons/lib/md'
import {Enum} from 'enumify'
import './NavBar.css'

export class ViewMode extends Enum {
  toJSON () {
    return this.name
  }
}

ViewMode.initEnum([
  'VIEW_GRAPH',
  'PAN_N_ZOOM',
  'EDIT_NODES',
  'EDIT_EDGES',
  'CREATE_NEW'
])

ViewMode.isEdgeMode = function (state) {
  return state.viewMode === ViewMode.EDIT_EDGES
}

ViewMode.isNodeMode = function (state) {
  return state.viewMode === ViewMode.EDIT_NODES
}

const VIEW_BUTTONS = [
  [ViewMode.VIEW_GRAPH, 'petc-vmc-view', '#view', MdRemoveRedEye],
  [ViewMode.PAN_N_ZOOM, 'petc-vmc-zpan', '#zpan', MdGames],
  [ViewMode.EDIT_NODES, 'petc-vmc-node', '#node', MdDeviceHub],
  [ViewMode.EDIT_EDGES, 'petc-vmc-edge', '#edge', MdShuffle],
  [ViewMode.CREATE_NEW, 'petc-vmc-demo', '/', MdCreateNewFolder]
]

export function viewModeFromHash (locationHash) {
  switch (locationHash) {
    case '#edge':
      return ViewMode.EDIT_EDGES
    case '#node':
      return ViewMode.EDIT_NODES
    case '#zpan':
      return ViewMode.PAN_N_ZOOM
    default:
      return ViewMode.VIEW_GRAPH
  }
}

export default class NavBar extends React.Component {
  getClass (mode) {
    return mode === this.props.viewMode ? 'active' : ''
  }

  render () {
    const buttons = VIEW_BUTTONS.map(([mode, buttonId, href, icon]) =>
      <li key={mode.toString()} className={this.getClass(mode)}>
        <a id={buttonId} href={href}>{React.createElement(icon)}</a>
      </li>
    )
    return (
      <div id='petc-view-mode-control'>
        <ul> {buttons} </ul>
      </div>
    )
  }
}
