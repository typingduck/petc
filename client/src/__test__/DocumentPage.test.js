/* global expect it jest */
import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import DocumentPage from '../DocumentPage'
import {ViewMode} from '../controls/NavBar'

jest.mock('../model/model', () => ({
  initialDoc: () => ({ nodes: {}, edges: {} }),
  loadDoc: () => Promise.resolve({ nodes: {}, edges: {} }),
  connectSocket: () => Promise.resolve()
}))

it('matches snapshot', () => {
  const props = {
    docId: '123',
    viewMode: ViewMode.VIEW_GRAPH,
    controls: { storeview: true },
    doc: { nodes: {}, edges: {} },
    docLoaded: () => {}
  }
  const renderer = new ShallowRenderer()
  const result = renderer.render(<DocumentPage {...props} />)
  expect(result).toMatchSnapshot()
})
