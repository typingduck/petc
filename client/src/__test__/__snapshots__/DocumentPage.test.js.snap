// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <NavBar
    controls={
      Object {
        "storeview": true,
      }
    }
    doc={
      Object {
        "edges": Object {},
        "nodes": Object {},
      }
    }
    docId="123"
    docLoaded={[Function]}
    viewMode={"VIEW_GRAPH"}
  />
  <div
    className="petc-pageview"
  >
    <h1>
      pannus et circulos
    </h1>
    <Kanvas
      addEdge={[Function]}
      addNode={[Function]}
      controls={
        Object {
          "storeview": true,
        }
      }
      doc={
        Object {
          "edges": Object {},
          "nodes": Object {},
        }
      }
      docId="123"
      docLoaded={[Function]}
      updateNode={[Function]}
      viewMode={"VIEW_GRAPH"}
    />
  </div>
  <StoreView
    controls={
      Object {
        "storeview": true,
      }
    }
    doc={
      Object {
        "edges": Object {},
        "nodes": Object {},
      }
    }
    docId="123"
    docLoaded={[Function]}
    viewMode={"VIEW_GRAPH"}
  />
</div>
`;
